java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+
+import java
+import semmle.code.java.controlflow.Guards
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.DataFlow3
+import semmle.code.java.dataflow.DataFlow4
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+import semmle.code.java.environment.SystemProperty
+import semmle.code.java.security.TempFileLib
+import DataFlow::PathGraph
+
+/**
+ * The flow state after the call to `System.getProperty`.
+ */
+private class FromSystemPropertyFlowState extends DataFlow::FlowState {
+  FromSystemPropertyFlowState() { this = "FromSystemPropertyFlowState" }
+}
+
+/**
+ * The flow state after the `File.createTempFile` method has been called on the tracked taint.
+ *
+ * This state may be achieved two ways:
+ *  1. If the source of the taint is a `File.createTempFile(_, _)` call.
+ *  2. From state `FromSystemPropertyFlowState` if `File.createTempFile(_, _, taint)` is called.
+ */
+private class FromFileCreateTempFileFlowState extends DataFlow::FlowState {
+  FromFileCreateTempFileFlowState() { this = "FromFileCreateTempFileFlowState" }
+}
+
+/**
+ * The flow state after the `File::delete` method has been called on tracked taint.
+ * This is the final state of the flow before hijackable call `File::mkdir(s)` (ie. the sink) is reached.
+ */
+private class FromDeleteFileFlowState extends DataFlow::FlowState {
+  FromDeleteFileFlowState() { this = "FromDeleteFileFlowState" }
+}
+
+/**
+ * The flow state after the `File::mkdir(s)` method has been called on tracked taint.
+ * This is the final state, and is only used when computing the 'unsafe' call site.
+ */
+private class FromMkdirsFileFlowState extends DataFlow::FlowState {
+  FromMkdirsFileFlowState() { this = "FromMkdirsFileFlowState" }
+}
+
+/**
+ * A guard that is checking if a directory exists, throwing if it does not exist.
+ */
+private Guard directoryExistsGuard(ThrowStmt s) {
+  result =
+    any(MethodAccess existanceCheck |
+      (
+        existanceCheck.getMethod() instanceof MethodFileExists
+        or
+        existanceCheck.getMethod() instanceof MethodFileIsDirectory
+      )
+    ) and
+    result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is verifying that the directory is sucsessfully created, throwing when it is not created.
+ */
+private Guard directoryCreationGuard(ThrowStmt s) {
+  result = any(MethodAccess creationCheck |
+    creationCheck.getMethod() instanceof MethodFileCreatesDirs
+  ) and
+  result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is safely verfying that a directory is created.
+ * 
+ * 'Safe' means that the directory is guarnteed to have been created, and the directory did not already exist.
+ */
+private Guard safeDirectoryCreationGuard(ThrowStmt s) {
+  result = directoryCreationGuard(s) and
+  // This guard is not 'safe' if a `directoryExistsGuard` is also guarding this throw statement
+  not exists(directoryExistsGuard(s))
+}
```

Comment:
There are a few additional cases that I want to test this predicate for, in particular, something like:

```java
void createIfNotPresent(File file) throws IOException {
    if (file.exists()) return;
    if (!file.mkdirs()) throw new IOException("Failed to make directory");
}

Reply:
Done. This predicate does cover that case correctly.